lib/SILOptimizer/Mandatory/YieldOnceCheck.cpp

```diff
@@ -0,0 +1,313 @@
+      return oldState;
+    }
+
+    if (newState.yieldState == BBState::Conflict) {
+      return newState;
+    }
+
+    if (oldState.yieldState == newState.yieldState) {
+      return oldState;
+    }
+
+    // Here, one state is AfterYield and the other one is BeforeYield.
+    // Merging them will result in Conflict.
+    SILInstruction *yieldInst = (newState.yieldState == BBState::AfterYield)
+                                    ? newState.getYieldInstruction()
+                                    : oldState.getYieldInstruction();
+
+    return BBState::getConflictState(yieldInst);
+  }
+
+  /// Perform a data-flow analysis to check whether there is exactly one
+  /// yield before a return in every path in the control-flow graph.
+  /// Diagnostics are not reported for nodes unreachable from the entry, and
+  /// also for nodes that may not reach the exit of the control-flow graph.
+  void diagnoseYieldOnceUsage(SILFunction &fun) {
+    llvm::DenseMap<SILBasicBlock *, BBState> bbToStateMap;
+    SmallVector<SILBasicBlock *, 16> worklist;
+
+    auto *entryBB = fun.getEntryBlock();
+    bbToStateMap.try_emplace(entryBB, BBState::getInitialState());
+    worklist.push_back(entryBB);
+
+    // ReturnBeforeYield errors, which denote that no paths yield before
+    // returning, are not diagnosed until the analysis completes, in order to
+    // distinguish them from ReturnOnConflict errors, which happen when some
+    // paths yield and some don't.
+    llvm::Optional<YieldError> returnBeforeYieldError = None;
+
+    // The algorithm uses a worklist to propagate the state through basic
+    // blocks until a fix point. Since the state lattice has height one, each
+    // basic block will be visited at most twice, and at most once if there are
+    // no Conflicts (which are errors). The basic blocks are added to the
+    // worklist in a breadth-first fashion. The order of visiting basic blocks
+    // is not important for correctness, but it could change the errors
+    // diagnosed when there are multiple errors. Breadth-first search diagnoses
+    // errors along shorter paths.
+    while (!worklist.empty()) {
+      SILBasicBlock *bb = worklist.pop_back_val();
+      const BBState &state = bbToStateMap.find(bb)->getSecond();
+
+      llvm::Optional<YieldError> errorResult = None;
+      auto resultState = transferStateThroughBasicBlock(bb, state, errorResult);
+
+      if (!resultState.hasValue()) {
+        auto error = errorResult.getValue();
+
+        // ReturnBeforeYield errors will not be reported until the analysis
+        // completes. So record it and continue.
+        if (error.errorKind == YieldError::ReturnBeforeYield) {
+          if (!returnBeforeYieldError.hasValue()) {
+            returnBeforeYieldError = error;
+          }
+          continue;
+        }
+
+        emitDiagnostics(error, fun);
+        return;
+      }
+
+      auto nextState = resultState.getValue();
+
+      for (auto &succ : bb->getSuccessors()) {
```

Comment:
Do you ever actually need access to the successor? If not you can use:

```
for (auto *succBB : bb->getSuccessorBlocks()) {
```

Reply:
Sure. Will do.